includes/class-event.php

```diff
@@ -0,0 +1,399 @@
+	|--------------------------------------------------------------------------
+	*/
+
+	public function set_status( string $status ): void {
+		$status = strtolower( $status );
+
+		if ( in_array( $status, Events_Store::ALLOWED_STATUSES, true ) ) {
+			$this->status = $status;
+		} else {
+			$this->status = Events_Store::STATUS_PENDING;
+		}
+
+		// Always mark status as changed, as we want to ensure it always sends this data in the SQL if needed.
+		$this->mark_changed( 'status' );
+	}
+
+	public function set_action( string $action ): void {
+		if ( ! empty( $action ) ) {
+			$this->action = $action;
+			$this->mark_changed( 'action' );
+
+			$this->action_hashed = md5( $action );
+			$this->mark_changed( 'action_hashed' );
+		}
+	}
+
+	public function set_args( array $args ): void {
+		$this->args = $args;
+		$this->mark_changed( 'args' );
+
+		$this->instance = self::create_instance_hash( $this->args );
+		$this->mark_changed( 'instance' );
+	}
+
+	public function set_schedule( string $schedule, int $interval ): void {
+		if ( ! empty( $schedule ) && $interval > 0 ) {
+			// Ensure either both or none are set.
+			$this->schedule = $schedule;
+			$this->mark_changed( 'schedule' );
+
+			$this->interval = $interval;
+			$this->mark_changed( 'interval' );
+		}
+	}
+
+	public function set_timestamp( int $timestamp ): void {
+		if ( $timestamp >= 1 ) {
+			$this->timestamp = $timestamp;
+			$this->mark_changed( 'timestamp' );
+		}
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Methods for interacting with the object.
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Save the event based on locally-changed props.
+	 *
+	 * @return true|WP_Error true on success, WP_Error on failure.
+	 */
+	public function save() {
+		$changed_data = $this->get_changed_data();
+
+		if ( empty( $changed_data ) ) {
+			// Nothing to save.
+			return new WP_Error( 'cron-control:event:no-save-needed' );
+		}
+
+		if ( ! isset( $this->action, $this->timestamp ) ) {
+			// These are required properties, must be set in order to save.
+			return new WP_Error( 'cron-control:event:missing-props' );
+		}
+
+		$all_row_data = [
+			'status'        => isset( $this->status ) ? $this->status : null,
+			'action'        => $this->action,
+			'action_hashed' => md5( $this->action ),
+			'args'          => serialize( $this->args ),
+			'instance'      => self::create_instance_hash( $this->args ),
+			'timestamp'     => $this->timestamp,
+		];
+
+		if ( isset( $this->schedule ) && isset( $this->interval ) ) {
+			$all_row_data['schedule'] = $this->schedule;
+			$all_row_data['interval'] = $this->interval;
+		}
+
+		// Pick out just the data that has changed.
+		$row_data = [];
+		foreach ( $changed_data as $changed_prop ) {
+			$row_data[ $changed_prop ] = $all_row_data[ $changed_prop ];
+		}
+
+		if ( $this->exists() ) {
+			$success = Events_Store::_update_event( $this->id, $row_data );
```

Comment:
I guess this goes somewhere else. But doesn't `_function_name` mean kinda internal as in private? As we call this outside of Events_Store it is ot private and I would guess the method should ot be underscored then. But I dont know much abut php naming convetntions to be honest.

Reply:
Yeah, the underscore is meant to showcase it as being preferable private (though unable to strictly enforce). Unfortunately the data store is a publicly accessible singleton atm though, so doing a "fake private" initially felt best until it's able to be made truly enclosed.

Reply:
Feels more like a subjective preference so sounds good to me as it is.